File: implementations/typescript/src/eval/positional.ts
import { Board } from "../board";
import { Color, PieceType, Square } from "../types";

const BISHOP_PAIR_BONUS = 30;
const ROOK_OPEN_FILE_BONUS = 25;
const ROOK_SEMI_OPEN_FILE_BONUS = 15;
const ROOK_SEVENTH_RANK_BONUS = 20;
const KNIGHT_OUTPOST_BONUS = 20;

export function evaluate(board: Board): number {
  let score = 0;
  
  score += evaluateColor(board, "white");
  score -= evaluateColor(board, "black");
  
  return score;
}

function evaluateColor(board: Board, color: Color): number {
  let score = 0;
  
  if (hasBishopPair(board, color)) {
    score += BISHOP_PAIR_BONUS;
  }
  
  for (let square = 0; square < 64; square++) {
    const piece = board.getPiece(square);
    if (piece && piece.color === color) {
      switch (piece.type) {
        case "R":
          score += evaluateRook(board, square, color);
          break;
        case "N":
          score += evaluateKnight(board, square, color);
          break;
      }
    }
  }
  
  return score;
}

function hasBishopPair(board: Board, color: Color): boolean {
  let bishopCount = 0;
  
  for (let square = 0; square < 64; square++) {
    const piece = board.getPiece(square);
    if (piece && piece.color === color && piece.type === "B") {
      bishopCount++;
    }
  }
  
  return bishopCount >= 2;
}

function evaluateRook(board: Board, square: Square, color: Color): number {
  const file = square % 8;
  const rank = Math.floor(square / 8);
  let bonus = 0;
  
  const [ownPawns, enemyPawns] = countPawnsOnFile(board, file, color);
  
  if (ownPawns === 0 && enemyPawns === 0) {
    bonus += ROOK_OPEN_FILE_BONUS;
  } else if (ownPawns === 0) {
    bonus += ROOK_SEMI_OPEN_FILE_BONUS;
  }
  
  const seventhRank = color === "white" ? 6 : 1;
  if (rank === seventhRank) {
    bonus += ROOK_SEVENTH_RANK_BONUS;
  }
  
  return bonus;
}

function evaluateKnight(board: Board, square: Square, color: Color): number {
  if (isOutpost(board, square, color)) {
    return KNIGHT_OUTPOST_BONUS;
  } else {
    return 0;
  }
}

function isOutpost(board: Board, square: Square, color: Color): boolean {
  const file = square % 8;
  const rank = Math.floor(square / 8);
  
  const protectedByPawn = isProtectedByPawn(board, square, color);
  if (!protectedByPawn) {
    return false;
  }
  
  const cannotBeAttacked = !canBeAttackedByEnemyPawn(board, square, file, rank, color);
  
  return protectedByPawn && cannotBeAttacked;
}

function isProtectedByPawn(board: Board, square: Square, color: Color): boolean {
  const file = square % 8;
  const rank = Math.floor(square / 8);
  
  const behindRank = color === "white" ? Math.max(0, rank - 1) : Math.min(7, rank + 1);
  
  const adjacentFiles = [Math.max(0, file - 1), Math.min(7, file + 1)];
  for (const adjacentFile of adjacentFiles) {
    if (adjacentFile !== file) {
      const checkSquare = behindRank * 8 + adjacentFile;
      const piece = board.getPiece(checkSquare);
      if (piece && piece.color === color && piece.type === "P") {
        return true;
      }
    }
  }
  
  return false;
}

function canBeAttackedByEnemyPawn(board: Board, square: Square, file: number, rank: number, color: Color): boolean {
  const startRank = color === "white" ? rank + 1 : 0;
  const endRank = color === "white" ? 8 : rank;
  
  for (let checkRank = startRank; color === "white" ? checkRank < endRank : checkRank < endRank; checkRank++) {
    const adjacentFiles = [Math.max(0, file - 1), Math.min(7, file + 1)];
    for (const adjacentFile of adjacentFiles) {
      if (adjacentFile !== file) {
        const checkSquare = checkRank * 8 + adjacentFile;
        const piece = board.getPiece(checkSquare);
        if (piece && piece.color !== color && piece.type === "P") {
          return true;
        }
      }
    }
  }
  
  return false;
}

function countPawnsOnFile(board: Board, file: number, color: Color): [number, number] {
  let ownPawns = 0;
  let enemyPawns = 0;
  
  for (let rank = 0; rank < 8; rank++) {
    const square = rank * 8 + file;
    const piece = board.getPiece(square);
    if (piece && piece.type === "P") {
      if (piece.color === color) {
        ownPawns++;
      } else {
        enemyPawns++;
      }
    }
  }
  
  return [ownPawns, enemyPawns];
}

File: implementations/rust/src/eval/positional.rs
use crate::board::Board;
use crate::types::{Color, PieceType, Square};

const BISHOP_PAIR_BONUS: i32 = 30;
const ROOK_OPEN_FILE_BONUS: i32 = 25;
const ROOK_SEMI_OPEN_FILE_BONUS: i32 = 15;
const ROOK_SEVENTH_RANK_BONUS: i32 = 20;
const KNIGHT_OUTPOST_BONUS: i32 = 20;

pub fn evaluate(board: &Board) -> i32 {
    let mut score = 0;
    
    score += evaluate_color(board, Color::White);
    score -= evaluate_color(board, Color::Black);
    
    score
}

fn evaluate_color(board: &Board, color: Color) -> i32 {
    let mut score = 0;
    
    if has_bishop_pair(board, color) {
        score += BISHOP_PAIR_BONUS;
    }
    
    for square in 0..64 {
        if let Some(piece) = board.get_piece(square) {
            if piece.color == color {
                match piece.piece_type {
                    PieceType::Rook => {
                        score += evaluate_rook(board, square, color);
                    },
                    PieceType::Knight => {
                        score += evaluate_knight(board, square, color);
                    },
                    _ => {},
                }
            }
        }
    }
    
    score
}

fn has_bishop_pair(board: &Board, color: Color) -> bool {
    let mut bishop_count = 0;
    
    for square in 0..64 {
        if let Some(piece) = board.get_piece(square) {
            if piece.color == color && piece.piece_type == PieceType::Bishop {
                bishop_count += 1;
            }
        }
    }
    
    bishop_count >= 2
}

fn evaluate_rook(board: &Board, square: Square, color: Color) -> i32 {
    let file = square % 8;
    let rank = square / 8;
    let mut bonus = 0;
    
    let (own_pawns, enemy_pawns) = count_pawns_on_file(board, file, color);
    
    if own_pawns == 0 && enemy_pawns == 0 {
        bonus += ROOK_OPEN_FILE_BONUS;
    } else if own_pawns == 0 {
        bonus += ROOK_SEMI_OPEN_FILE_BONUS;
    }
    
    let seventh_rank = if color == Color::White { 6 } else { 1 };
    if rank == seventh_rank {
        bonus += ROOK_SEVENTH_RANK_BONUS;
    }
    
    bonus
}

fn evaluate_knight(board: &Board, square: Square, color: Color) -> i32 {
    if is_outpost(board, square, color) {
        KNIGHT_OUTPOST_BONUS
    } else {
        0
    }
}

fn is_outpost(board: &Board, square: Square, color: Color) -> bool {
    let file = square % 8;
    let rank = square / 8;
    
    let protected_by_pawn = is_protected_by_pawn(board, square, color);
    if !protected_by_pawn {
        return false;
    }
    
    let cannot_be_attacked = !can_be_attacked_by_enemy_pawn(board, square, file, rank, color);
    
    protected_by_pawn && cannot_be_attacked
}

fn is_protected_by_pawn(board: &Board, square: Square, color: Color) -> bool {
    let file = square % 8;
    let rank = square / 8;
    
    let behind_rank = if color == Color::White {
        rank.saturating_sub(1)
    } else {
        (rank + 1).min(7)
    };
    
    for adjacent_file in [file.saturating_sub(1), (file + 1).min(7)].iter() {
        if *adjacent_file != file {
            let check_square = behind_rank * 8 + adjacent_file;
            if let Some(piece) = board.get_piece(check_square) {
                if piece.color == color && piece.piece_type == PieceType::Pawn {
                    return true;
                }
            }
        }
    }
    
    false
}

fn can_be_attacked_by_enemy_pawn(board: &Board, square: Square, file: Square, rank: Square, color: Color) -> bool {
    let ahead_ranks = if color == Color::White {
        rank + 1..8
    } else {
        0..rank
    };
    
    for check_rank in ahead_ranks {
        for adjacent_file in [file.saturating_sub(1), (file + 1).min(7)].iter() {
            if *adjacent_file != file {
                let check_square = check_rank * 8 + adjacent_file;
                if let Some(piece) = board.get_piece(check_square) {
                    if piece.color != color && piece.piece_type == PieceType::Pawn {
                        return true;
                    }
                }
            }
        }
    }
    
    false
}

fn count_pawns_on_file(board: &Board, file: Square, color: Color) -> (i32, i32) {
    let mut own_pawns = 0;
    let mut enemy_pawns = 0;
    
    for rank in 0..8 {
        let square = rank * 8 + file;
        if let Some(piece) = board.get_piece(square) {
            if piece.piece_type == PieceType::Pawn {
                if piece.color == color {
                    own_pawns += 1;
                } else {
                    enemy_pawns += 1;
                }
            }
        }
    }
    
    (own_pawns, enemy_pawns)
}
